src/aggregator/client/writer.go

```diff
@@ -147,10 +147,19 @@ func (w *writer) Close() error {
 		return errInstanceWriterClosed
 	}
 	w.closed = true
-	if err := w.flushWithLock(); err != nil {
-		w.metrics.flushErrors.Inc(1)
-	}
-	return w.queue.Close()
+
+	go func() {
+		if err := w.flushWithLock(); err != nil {
```

Comment:
Hm you no longer hold the lock in this goroutine.  I think you might want to `w.Lock()` and `defer w.Unlock()` from inside the async goroutine too?

Reply:
Either that or conditionally unlock, and if you get to this goroutine, `defer w.Unlock()` from the top of the goroutine execution start?

Reply:
I assume that `w.queue.Close()` you can safely do from an async goroutine, but `flushWithLock()` it doesn't look like it.

Maybe you can rename `flushWithLock()` to `flushWithEncodersAndQueue(...)` take the params it works on as arguments then nil them out during close and pass them yourself async from the goroutine?

Reply:
flushWithLock should also be safe because `w.closed` was set to true, so no other writes will be accepted after that, but Richie's suggestion seems cleaner to do it asynchronously in `mgr.RemoveInstance()`, I'll go with that route instead